esp/esp/program/modules/base.py

```diff
@@ -512,28 +512,56 @@ def _checkGrade(moduleObj, request, tl, *args, **kwargs):
 
 # Just broke out this function to allow combined deadlines (see meets_any_deadline,
 # meets_all_deadlines functions below).  -Michael P, 6/23/2009
+
+# Combined deadlines are annoying; meets_deadline has been absorbed into
+# meets_any_deadline and meets_all_deadlines has been deleted --bpchen,
+# 2017-02-17
 def _checkDeadline_helper(method, extension, moduleObj, request, tl, *args, **kwargs):
+    """
+    Given information about a request, return a pair of type (bool, None |
+    response), which indicates whether the user can view the requested page,
+    and an optional redirect if not.
+
+    If the user is an administrator, annotate the request with information
+    about what roles have permission to view the requested page.
+    """
     if tl != 'learn' and tl != 'teach' and tl != 'volunteer':
         return (True, None)
     response = None
     canView = False
     perm_name = {'learn':'Student','teach':'Teacher','volunteer':'Volunteer'}[tl]+extension
     if not_logged_in(request):
-        if not moduleObj.require_auth() and Permission.valid_objects().filter(permission_type=perm_name, program=request.program, user__isnull=True).exists():
+        if not moduleObj.require_auth() and Permission.null_user_has_perm(permission_type=perm_name, program=request.program):
             canView = True
         else:
             response = HttpResponseRedirect('%s?%s=%s' % (LOGIN_URL, REDIRECT_FIELD_NAME, quote(request.get_full_path())))
     else:
-        canView = request.user.updateOnsite(request)
+        user = request.user
+        program = request.program
+        canView = user.updateOnsite(request)
         if not canView:
-            canView = Permission.user_has_perm(request.user,
+            canView = Permission.user_has_perm(user,
                                                perm_name,
-                                               program=request.program)
+                                               program=program)
             #   For now, allow an exception if the user is of the wrong type
             #   This is because we are used to UserBits having a null user affecting everyone, regardless of user type.
-            if not canView and Permission.valid_objects().filter(permission_type=perm_name, program=request.program, user__isnull=True).exists():
+            if not canView and Permission.valid_objects().filter(permission_type=perm_name, program=program, user__isnull=True).exists():
                 canView = True
 
+            #   Give administrators additional information
+            if user.isAdministrator(program=program):
+                request.show_perm_info = True
+                if hasattr(request, 'perm_names'):
```

Comment:
[`hasattr` has some weird edge cases](https://hynek.me/articles/hasattr/); normally I don't worry too much about them, but request objects in django are super duper magic so it could actually be an issue.  Better to use `getattr(obj, attr, None) is not None`.